File: scripts/wandb_utils.py
import os
import json
import re
import yaml
from datetime import datetime, timezone
from pathlib import Path


def load_json_file(file_path):
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as e:
        print(f"✗ Error loading {file_path}: {e}")
        return None


def load_yaml_file(file_path):
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            return yaml.safe_load(f)
    except ImportError:
        print("PyYAML not found. Please install it using: pip install PyYAML")
        return None
    except Exception as e:
        print(f"Error loading {file_path}: {e}")
        return None


def is_integer_folder(folder_name):
    try:
        int(folder_name)
        return True
    except ValueError:
        return False


def find_json_files(integer_folder_path):
    performance_file = None
    result_file = None
    
    for file in os.listdir(integer_folder_path):
        if file.endswith('_performance.json'):
            performance_file = os.path.join(integer_folder_path, file)
        elif file.endswith('_result.json'):
            result_file = os.path.join(integer_folder_path, file)
    
    return performance_file, result_file


def find_config_file(subfolder_path):
    config_path = os.path.join(subfolder_path, "config.yaml")
    if os.path.exists(config_path):
        return config_path
    return None


def find_wandb_metadata_file(subfolder_path):
    metadata_path = os.path.join(subfolder_path, "wandb-metadata.json")
    if os.path.exists(metadata_path):
        return metadata_path
    return None


def parse_llm_string_with_eval(llm_str):
    try:
        cleaned_str = re.sub(r'\$\{oc\.env:[^}]+\}', '""', llm_str)
        cleaned_str = re.sub(r'\$\{[^}]+\}', '""', cleaned_str)        
        result = eval(cleaned_str)
        return result
    except Exception as e:
        return None


def extract_model_config(model, llm_config):
    litellm_params = model.get('litellm_params', {})
    completion_params = llm_config.get('completion_params', {})
    
    model_config = {
        "model_name": model.get('model_name', ''),
        "model": litellm_params.get('model', ''),
        "api_base": litellm_params.get('api_base', ''),
        "timeout": litellm_params.get('timeout', 300),
        "num_retries": litellm_params.get('num_retries', 5),
        "rpm": litellm_params.get('rpm', None),
        "top_p": completion_params.get('top_p', None),
        "max_tokens": completion_params.get('max_tokens', None),
        "temperature": completion_params.get('temperature', None)
    }
    
    model_config = {k: v for k, v in model_config.items() if v is not None}
    return model_config


def parse_llm_configs(config_data):
    llm_configs = []
    llm_keys = ['llm1', 'llm2', 'llm3']
    
    for llm_key in llm_keys:
        if llm_key in config_data:
            llm_data = config_data[llm_key]
            
            if isinstance(llm_data, dict) and 'value' in llm_data:
                llm_str = llm_data['value']
            else:
                continue
            
            try:
                llm_config = parse_llm_string_with_eval(llm_str)
                
                if llm_config and isinstance(llm_config, dict) and 'language_models' in llm_config:
                    for model in llm_config['language_models']:
                        model_config = extract_model_config(model, llm_config)
                        llm_configs.append(model_config)
                
            except Exception as e:
                print(f"⚠ Error parsing {llm_key} config: {e}")
    
    return llm_configs


def get_llm_config_key(llm_config):
    return f"{llm_config.get('model_name', '')}|{llm_config.get('model', '')}"


def remove_duplicate_llm_configs(llm_configs):
    seen_configs = set()
    unique_configs = []
    
    for config in llm_configs:
        config_key = get_llm_config_key(config)
        if config_key not in seen_configs:
            seen_configs.add(config_key)
            unique_configs.append(config)
    
    return unique_configs


def process_wandb_metadata(metadata_data):
    if not metadata_data:
        return {}
    
    processed_metadata = {
        "startedAt": metadata_data.get("startedAt", ""),
    }
    
    args_dict = {}
    if "args" in metadata_data and isinstance(metadata_data["args"], list):
        for arg in metadata_data["args"]:
            if "=" in arg:
                key, value = arg.split("=", 1)
                try:
                    if "." in value:
                        value = float(value)
                    else:
                        value = int(value)
                except ValueError:
                    if value.lower() in ["true", "false"]:
                        value = value.lower() == "true"
                args_dict[key] = value
    
    processed_metadata["parsed_args"] = args_dict
    
    return processed_metadata


def process_wandb_data(wandb_data_path):
    results = []
    
    if not os.path.exists(wandb_data_path):
        print(f"Path {wandb_data_path} does not exist")
        return results
    
    total_processed = 0
    
    for subfolder in os.listdir(wandb_data_path):
        subfolder_path = os.path.join(wandb_data_path, subfolder)
        
        if not os.path.isdir(subfolder_path):
            continue
                
        config_file = find_config_file(subfolder_path)
        config_data = None
        llm_configs = []
        
        if config_file:
            config_data = load_yaml_file(config_file)
            if config_data:
                llm_configs = parse_llm_configs(config_data)
                llm_configs = remove_duplicate_llm_configs(llm_configs)
        
        metadata_file = find_wandb_metadata_file(subfolder_path)
        metadata_data = None
        processed_metadata = {}
        
        if metadata_file:
            metadata_data = load_json_file(metadata_file)
            if metadata_data:
                processed_metadata = process_wandb_metadata(metadata_data)
        
        for item in os.listdir(subfolder_path):
            item_path = os.path.join(subfolder_path, item)
            
            if os.path.isdir(item_path) and is_integer_folder(item):
                performance_file, result_file = find_json_files(item_path)
                
                if not performance_file or not result_file:
                    continue
                
                performance_data = load_json_file(performance_file)
                result_data = load_json_file(result_file)
                
                if performance_data is None or result_data is None:
                    continue
                
                combined_data = {
                    "status": "completed",
                    "performance_data": performance_data,
                    "result_data": result_data,
                    "modelConfig": {
                        "LLM": llm_configs,
                        "Human": []
                    },
                    "wandb_metadata": processed_metadata,
                    "processed_at": datetime.now(timezone.utc)
                }
                
                results.append(combined_data)
                total_processed += 1
    
    print(f"Processing {total_processed} records...")
    return results


def get_statistics(processed_data):
    if not processed_data:
        return {}
    
    all_llm_configs = []
    for data in processed_data:
        all_llm_configs.extend(data['modelConfig']['LLM'])
    
    unique_llm_configs = remove_duplicate_llm_configs(all_llm_configs)
    metadata_count = sum(1 for data in processed_data if data.get('wandb_metadata'))
    
    return {
        "total_records": len(processed_data),
        "unique_llm_configs": len(unique_llm_configs),
        "records_with_metadata": metadata_count
    }

File: backend/src/scripts/wandb_utils.js
const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

function loadJsonFile(filePath) {
  try {
    const data = fs.readFileSync(filePath, 'utf8');
    return JSON.parse(data);
  } catch (error) {
    console.log(`✗ Error loading ${filePath}: ${error.message}`);
    return null;
  }
}

function loadYamlFile(filePath) {
  try {
    const data = fs.readFileSync(filePath, 'utf8');
    return yaml.load(data);
  } catch (error) {
    if (error.code === 'MODULE_NOT_FOUND') {
      console.log('js-yaml not found. Please install it using: npm install js-yaml');
    } else {
      console.log(`Error loading ${filePath}: ${error.message}`);
    }
    return null;
  }
}

function isIntegerFolder(folderName) {
  const num = parseInt(folderName, 10);
  return !isNaN(num) && num.toString() === folderName;
}

function findJsonFiles(integerFolderPath) {
  let performanceFile = null;
  let resultFile = null;

  try {
    const files = fs.readdirSync(integerFolderPath);
    
    for (const file of files) {
      if (file.endsWith('_performance.json')) {
        performanceFile = path.join(integerFolderPath, file);
      } else if (file.endsWith('_result.json')) {
        resultFile = path.join(integerFolderPath, file);
      }
    }
  } catch (error) {
    console.log(`Error reading directory ${integerFolderPath}: ${error.message}`);
  }

  return { performanceFile, resultFile };
}

function findConfigFile(subfolderPath) {
  const configPath = path.join(subfolderPath, 'config.yaml');
  return fs.existsSync(configPath) ? configPath : null;
}

function findWandbMetadataFile(subfolderPath) {
  const metadataPath = path.join(subfolderPath, 'wandb-metadata.json');
  return fs.existsSync(metadataPath) ? metadataPath : null;
}

function parseLlmStringWithEval(llmStr) {
  try {
    let cleanedStr = llmStr.replace(/\$\{oc\.env:[^}]+\}/g, '""');
    cleanedStr = cleanedStr.replace(/\$\{[^}]+\}/g, '""');
    const result = eval(`(${cleanedStr})`);
    return result;
  } catch (error) {
    console.log(`Error parsing LLM string: ${error.message}`);
    return null;
  }
}

function extractModelConfig(model, llmConfig) {
  const litellmParams = model.litellm_params || {};
  const completionParams = llmConfig.completion_params || {};

  const modelConfig = {
    model_name: model.model_name || '',
    model: litellmParams.model || '',
    api_base: litellmParams.api_base || '',
    timeout: litellmParams.timeout || 300,
    num_retries: litellmParams.num_retries || 5,
    rpm: litellmParams.rpm || null,
    top_p: completionParams.top_p || null,
    max_tokens: completionParams.max_tokens || null,
    temperature: completionParams.temperature || null,
  };

  const filteredConfig = {};
  for (const [key, value] of Object.entries(modelConfig)) {
    if (value !== null) {
      filteredConfig[key] = value;
    }
  }

  return filteredConfig;
}

function extractSeedFromConfig(configData) {
  if (configData.seed && configData.seed.value !== undefined) {
    const seedValue = configData.seed.value;
    if (typeof seedValue === 'number') {
      return seedValue;
    } else if (typeof seedValue === 'string') {
      const parsedSeed = parseInt(seedValue, 10);
      if (!isNaN(parsedSeed)) {
        return parsedSeed;
      }
    }
  }
  
  if (configData.seed !== undefined && typeof configData.seed === 'number') {
    return configData.seed;
  }
  
  if (configData.defaults && configData.defaults.seed && configData.defaults.seed.value !== undefined) {
    return configData.defaults.seed.value;
  }
  
  if (configData.hydra && configData.hydra.job && configData.hydra.job.override_dirname) {
    const overrideParts = configData.hydra.job.override_dirname.split(',');
    for (const part of overrideParts) {
      if (part.includes('seed=')) {
        const seedValue = part.split('seed=')[1];
        const parsedSeed = parseInt(seedValue, 10);
        if (!isNaN(parsedSeed)) {
          return parsedSeed;
        }
      }
    }
  }
  
  if (configData.run && configData.run.seed && configData.run.seed.value !== undefined) {
    return configData.run.seed.value;
  }
  
  if (configData.experiment && configData.experiment.seed && configData.experiment.seed.value !== undefined) {
    return configData.experiment.seed.value;
  }
  
  return null;
}

function extractDatasetFromConfig(configData) {
  if (configData.task && configData.task.value !== undefined) {
    const taskValue = configData.task.value;
    
    if (typeof taskValue === 'string') {
      try {
        let cleanedStr = taskValue.replace(/'/g, '"');
        const taskConfig = JSON.parse(cleanedStr);
        
        if (taskConfig.name) {
          return taskConfig.name;
        }
      } catch (error) {
        const nameMatch = taskValue.match(/'name':\s*'([^']+)'/);
        if (nameMatch && nameMatch[1]) {
          return nameMatch[1];
        }
      }
    }
    
    if (typeof taskValue === 'object' && taskValue !== null && taskValue.name) {
      return taskValue.name;
    }
  }
  
  if (configData.task && typeof configData.task === 'object' && configData.task.name) {
    return configData.task.name;
  }
  
  if (configData.defaults && configData.defaults.task && configData.defaults.task.value) {
    const taskValue = configData.defaults.task.value;
    if (typeof taskValue === 'string') {
      try {
        let cleanedStr = taskValue.replace(/'/g, '"');
        const taskConfig = JSON.parse(cleanedStr);
        if (taskConfig.name) {
          return taskConfig.name;
        }
      } catch (error) {
        const nameMatch = taskValue.match(/'name':\s*'([^']+)'/);
        if (nameMatch && nameMatch[1]) {
          return nameMatch[1];
        }
      }
    }
  }
  
  return null;
}

function parseLlmConfigs(configData) {
  const llmConfigs = [];
  const llmKeys = ['llm1', 'llm2', 'llm3'];

  for (const llmKey of llmKeys) {
    if (llmKey in configData) {
      const llmData = configData[llmKey];

      let llmStr;
      if (typeof llmData === 'object' && llmData !== null && 'value' in llmData) {
        llmStr = llmData.value;
      } else {
        continue;
      }

      try {
        const llmConfig = parseLlmStringWithEval(llmStr);

        if (llmConfig && 
            typeof llmConfig === 'object' && 
            'language_models' in llmConfig) {
          
          for (const model of llmConfig.language_models) {
            const modelConfig = extractModelConfig(model, llmConfig);
            llmConfigs.push(modelConfig);
          }
        }
      } catch (error) {
        console.log(`⚠ Error parsing ${llmKey} config: ${error.message}`);
      }
    }
  }

  return llmConfigs;
}

function getLlmConfigKey(llmConfig) {
  return `${llmConfig.model_name || ''}|${llmConfig.model || ''}`;
}

function removeDuplicateLlmConfigs(llmConfigs) {
  const seenConfigs = new Set();
  const uniqueConfigs = [];

  for (const config of llmConfigs) {
    const configKey = getLlmConfigKey(config);
    if (!seenConfigs.has(configKey)) {
      seenConfigs.add(configKey);
      uniqueConfigs.push(config);
    }
  }

  return uniqueConfigs;
}

function processWandbMetadata(metadataData) {
  if (!metadataData) {
    return {};
  }

  const processedMetadata = {
    startedAt: metadataData.startedAt || '',
  };

  const argsDict = {};
  if (metadataData.args && Array.isArray(metadataData.args)) {
    for (const arg of metadataData.args) {
      if (arg.includes('=')) {
        const [key, ...valueParts] = arg.split('=');
        let value = valueParts.join('=');

        if (/^\d+$/.test(value)) {
          value = parseInt(value, 10);
        } else if (/^\d+\.\d+$/.test(value)) {
          value = parseFloat(value);
        } else if (value.toLowerCase() === 'true') {
          value = true;
        } else if (value.toLowerCase() === 'false') {
          value = false;
        }
        argsDict[key] = value;
      }
    }
  }

  processedMetadata.parsed_args = argsDict;
  return processedMetadata;
}

function processWandbData(wandbDataPath) {
  const results = [];

  if (!fs.existsSync(wandbDataPath)) {
    console.log(`Path ${wandbDataPath} does not exist`);
    return results;
  }

  let totalProcessed = 0;

  try {
    const subfolders = fs.readdirSync(wandbDataPath);

    for (const subfolder of subfolders) {
      const subfolderPath = path.join(wandbDataPath, subfolder);

      if (!fs.statSync(subfolderPath).isDirectory()) {
        continue;
      }

      const configFile = findConfigFile(subfolderPath);
      let configData = null;
      let llmConfigs = [];
      let currentSeed = null;
      let datasetName = null;

      if (configFile) {
        configData = loadYamlFile(configFile);
        if (configData) {
          llmConfigs = parseLlmConfigs(configData);
          llmConfigs = removeDuplicateLlmConfigs(llmConfigs);
          currentSeed = extractSeedFromConfig(configData);
          datasetName = extractDatasetFromConfig(configData);
        }
      }

      const metadataFile = findWandbMetadataFile(subfolderPath);
      let metadataData = null;
      let processedMetadata = {};

      if (metadataFile) {
        metadataData = loadJsonFile(metadataFile);
        if (metadataData) {
          processedMetadata = processWandbMetadata(metadataData);
        }
      }

      const items = fs.readdirSync(subfolderPath);

      for (const item of items) {
        const itemPath = path.join(subfolderPath, item);

        if (fs.statSync(itemPath).isDirectory() && isIntegerFolder(item)) {
          const { performanceFile, resultFile } = findJsonFiles(itemPath);

          if (!performanceFile || !resultFile) {
            continue;
          }

          const performanceData = loadJsonFile(performanceFile);
          const resultData = loadJsonFile(resultFile);

          if (performanceData === null || resultData === null) {
            continue;
          }

          const combinedData = {
            status: 'completed',
            performance_data: performanceData,
            result_data: resultData,
            modelConfig: {
              LLM: llmConfigs,
              Human: [],
            },
            wandb_metadata: processedMetadata,
            current_seed: currentSeed,
            dataset_name: datasetName,
            processed_at: new Date(),
          };

          results.push(combinedData);
          totalProcessed++;
        }
      }
    }
  } catch (error) {
    console.error(`Error processing wandb data: ${error.message}`);
  }

  console.log(`Processing ${totalProcessed} records...`);
  return results;
}

function getStatistics(processedData) {
  if (!processedData || processedData.length === 0) {
    return {};
  }

  const allLlmConfigs = [];
  const seedCounts = {};
  const datasetCounts = {};
  let recordsWithSeed = 0;
  let recordsWithDataset = 0;

  for (const data of processedData) {
    allLlmConfigs.push(...data.modelConfig.LLM);
    
    if (data.current_seed !== null) {
      recordsWithSeed++;
      const seed = data.current_seed.toString();
      seedCounts[seed] = (seedCounts[seed] || 0) + 1;
    }
    
    if (data.dataset_name !== null) {
      recordsWithDataset++;
      const dataset = data.dataset_name;
      datasetCounts[dataset] = (datasetCounts[dataset] || 0) + 1;
    }
  }

  const uniqueLlmConfigs = removeDuplicateLlmConfigs(allLlmConfigs);
  const metadataCount = processedData.filter(data => data.wandb_metadata).length;

  return {
    totalRecords: processedData.length,
    uniqueLlmConfigs: uniqueLlmConfigs.length,
    recordsWithMetadata: metadataCount,
    recordsWithSeed: recordsWithSeed,
    uniqueSeeds: Object.keys(seedCounts).length,
    seedDistribution: seedCounts,
    recordsWithDataset: recordsWithDataset,
    uniqueDatasets: Object.keys(datasetCounts).length,
    datasetDistribution: datasetCounts,
  };
}

module.exports = {
  loadJsonFile,
  loadYamlFile,
  isIntegerFolder,
  findJsonFiles,
  findConfigFile,
  findWandbMetadataFile,
  parseLlmStringWithEval,
  extractModelConfig,
  extractSeedFromConfig,
  extractDatasetFromConfig,
  parseLlmConfigs,
  getLlmConfigKey,
  removeDuplicateLlmConfigs,
  processWandbMetadata,
  processWandbData,
  getStatistics,
};